File: algos/380-randomizedSet/380-randomizedSet.js

var RandomizedSet = function() {
    // num: indexInData
    this.indexMap = {}
    this.data = []
    this.length = 0
};

/** 
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.insert = function(val) {
    if (this.indexMap.hasOwnProperty(val)) return false;

    // update array
    this.data.push(val);

    // update hashMap
    this.indexMap[val] = this.length;

    // update length
    this.length++;
    return true;
};

/** 
 * @param {number} val
 * @return {boolean}
 */
RandomizedSet.prototype.remove = function(val) {
    if (this.indexMap.hasOwnProperty(val)) {
        // update length
        this.length--;

        const valIndex = this.indexMap[val];
        const lastEl = this.data[this.length];

        // update array
        this.data[valIndex] = lastEl;
        this.data.pop();

        // update hashMap
        this.indexMap[lastEl] = valIndex;
        delete this.indexMap[val]

        return true;
    } 
    return false;
};

/**
 * @return {number}
 */
RandomizedSet.prototype.getRandom = function() {
    const randomIndex = Math.floor(Math.random()*this.length)
    return this.data[randomIndex]
};

/** 
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */

File: algos/380-randomizedSet/380-randomizedSet.py
class RandomizedSet:

    def __init__(self):
        self.numMap = {} # num: indexOfNumList
        self.numList = [] 

    def insert(self, val: int) -> bool:
        valExists = val in self.numMap

        if valExists:
            return False

        # update map
        self.numMap[val] = len(self.numList)

        # update list
        self.numList.append(val)

        return True

    def remove(self, val: int) -> bool:
        valExists = val in self.numMap

        if valExists:
            valIndex = self.numMap[val]
            lastVal = self.numList[len(self.numList) - 1]

            # update list
            self.numList[valIndex] = lastVal
            self.numList.pop()

            # update map
            self.numMap[lastVal] = valIndex
            del self.numMap[val]
            return True

        return False

    def getRandom(self) -> int:
        return random.choice(self.numList)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()